app/src/main/java/com/hamdan/forzenbook/data/repository/LoginRepositoryImpl.kt

```diff
@@ -1,22 +1,29 @@
 package com.hamdan.forzenbook.data.repository
 
 import com.hamdan.forzenbook.data.database.LoginDao
-import com.hamdan.forzenbook.data.network.LoginResponse
+import com.hamdan.forzenbook.data.database.LoginEntity
 import com.hamdan.forzenbook.data.network.LoginService
 
 class LoginRepositoryImpl(
-    // TODO implement the dao FA-84
     private val dao: LoginDao,
     private val service: LoginService
 ) : LoginRepository {
     override suspend fun requestValidation(email: String) {
         service.requestValidation(email).isSuccessful
     }
 
-    override suspend fun getToken(email: String, code: String): LoginResponse {
+    override suspend fun getToken(): LoginData? {
+        // we catch the exception up the chain if there is an issue
+        return dao.getToken()?.toLoginData()
+    }
+
+    override suspend fun getToken(email: String, code: String): LoginData {
```

Comment:
The repo should be deciding which data method to retrieve from. I think it's best to combine these methods and let the repo choose to grab from db or go to the network. Either that or separate their naming?

Reply:
I think seperating their naming is the more viable option, it feels really pointless to purposely send down an empty email and code just so an if can see it so it does a different action.